barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/sorted_msm.cpp

```diff
@@ -0,0 +1,240 @@
+#include "barretenberg/ecc/scalar_multiplication/sorted_msm.hpp"
+
+namespace bb {
+
+/**
+ * @brief Reduce MSM inputs such that the set of scalars contains no duplicates by summing points which share a scalar.
+ * @details Since point addition is substantially cheaper than scalar multiplication, it is more efficient in some cases
+ * to first sum all points which share a scalar then perform the MSM on the reduced set of inputs. This is achieved via
+ * the following procedure:
+ *
+ * 1) Sort the input {points, scalars} by scalar in order to group points into 'addition sequences' i.e. sets of points
+ * to be added together prior to performing the MSM.
+ *
+ * 2) For each sequence, perform pairwise addition on all points. (If the length of the sequence is odd, the unpaired
+ * point is simply carried over to the next round). The inverses needed in the addition formula are batch computed in a
+ * single go for all additions to be performed across all sequences in a given round.
+ *
+ * 3) Perform rounds of pair-wise addition until each sequence is reduced to a single point.
+ *
+ * @tparam Curve
+ * @param scalars
+ * @param points
+ * @return MsmSorter<Curve>::ReducedMsmInputs
+ */
+template <typename Curve>
+MsmSorter<Curve>::ReducedMsmInputs MsmSorter<Curve>::reduce_msm_inputs(std::span<Fr> scalars, std::span<G1> points)
+{
+    // Generate the addition sequences (sets of points sharing a scalar)
+    AdditionSequences addition_sequences = construct_addition_sequences(scalars, points);
+
+    // Perform rounds of pairwise addition until all sets of points sharing a scalar have been reduced to a single point
+    batched_affine_add_in_place(addition_sequences);
+
+    // The reduced MSM inputs are the unique scalars and the reduced points
+    std::span<Fr> output_scalars(unique_scalars.data(), num_unique_scalars);
+    std::span<G1> output_points(updated_points.data(), num_unique_scalars);
+    return { output_scalars, output_points };
+}
+
+/**
+ * @brief Sort the MSM points by scalar so that points sharing a scalar can be summed prior to performing MSM
+ *
+ * @tparam Curve
+ * @param scalars
+ * @param points
+ * @return MsmSorter<Curve>::AdditionSequences
+ */
+template <typename Curve>
+MsmSorter<Curve>::AdditionSequences MsmSorter<Curve>::construct_addition_sequences(std::span<Fr> scalars,
+                                                                                   std::span<G1> points)
+{
+    // Create the array containing the indices of the scalars and points sorted by scalar value
+    const size_t num_points = points.size();
+    std::iota(index.begin(), index.end(), 0);
```

Comment:
Probably would be good to parallelise both for larger ranges. For sort you can do similar to https://github.com/AztecProtocol/aztec-packages/blob/59b92ca0f72ca3705dd6933b304897c91edc81c3/barretenberg/cpp/src/barretenberg/honk/proof_system/permutation_library.hpp#L191

Reply:
Good call thanks. BTW is there a typo in the code you linked? Seems like in `else` it should be `extra_denominator_uint.end()` instead of `extra_denominator.end()`

Reply:
@Rumata888 I got the following error when running `./bootstrap.sh` in cpp ( on v0.46.1 ).
```
.../honk/proof_system/permutation_library.hpp:191:74: fatal error: use of undeclared identifier 'extra_denominator'; did you mean 'extra_denominator_uint'?
  191 |     std::sort(std::execution::par_unseq, extra_denominator_uint.begin(), extra_denominator.end());
      |                                                                          ^~~~~~~~~~~~~~~~~
      |                                                                          extra_denominator_uint
```

Reply:
Is it the same error on latest master?